core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchWriter.java

```diff
@@ -639,6 +590,39 @@ public MutationWriter(int numSendThreads) {
       binningThreadPool =
           ThreadPools.createFixedThreadPool(1, "BinMutations", new SynchronousQueue<>(), false);
       binningThreadPool.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+      binningThreadPool.execute(Trace.wrap(new Runnable() {
+        @Override
+        public void run() {
+          while (!closed) {
+            int num = incomingQueue.numMutations();
+            if (num > 0) {
+              try {
+                log.trace("{} - binning {} mutations", Thread.currentThread().getName(), num);
+                // Remove the mutations from the incoming queue
+                final MutationSet copy = new MutationSet();
+                incomingQueue.forEach((table, mutations) -> {
+                  mutations.forEach(mutation -> {
+                    copy.add(table, mutation);
+                    mutations.remove(mutation);
+                  });
```

Comment:
Calling `remove()` on a collection while iterating over it may lead to unpredictable results.  Could use poll() since this is a Queue or user an iterator and call remove on it.

```suggestion
                  Mutation m = null;
                  while((m = mutations.poll()) != null) {
                	      copy.add(table, m);
                  }
```

Reply:
I implemented your suggestion.